Intel HDA: dump registers and trace when they might have been valid

ujfalusi opened the pull request:
We have the `flags` passed for the dsp_dump and they are mostly ignored all around the code.

Instead of printing the full debug trace when the fw_state is SOF_FW_BOOT_COMPLETE, use the flags to decide to print full or extended debug dump.

For this to make sense we also need to drop the SOF_DBG_DUMP_REGS from few places to be passed to snd_sof_dsp_dbg_dump() where we do not expect to have any backtrace, like when we just powering up the DSP.

lgirdwood commented on sound/soc/sof/loader.c:
```diff
@@ -821,8 +821,7 @@ int snd_sof_run_firmware(struct snd_sof_dev *sdev)
 		dev_err(sdev->dev, "error: failed to start DSP\n");
-		snd_sof_dsp_dbg_dump(sdev, SOF_DBG_DUMP_REGS | SOF_DBG_DUMP_MBOX |
-				     SOF_DBG_DUMP_PCI);
+		snd_sof_dsp_dbg_dump(sdev, SOF_DBG_DUMP_MBOX | SOF_DBG_DUMP_PCI);
```
Why do we want to hide details in the event of a non FW boot. Wouldn't this be helpful in debugging ?

ujfalusi replied:
These _were_ hidden as well before and at this point the DSP has not started to boot afaik. Not much to report about.

ujfalusi replied:
Initially I only want to enable dumps from point where it certainly makes sense and we can expect valid information.

ranj063 commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -184,8 +184,7 @@ static int cl_dsp_init(struct snd_sof_dev *sdev, int stream_tag)
 			__func__);
 
 err:
-	flags = SOF_DBG_DUMP_REGS | SOF_DBG_DUMP_PCI | SOF_DBG_DUMP_MBOX |
```
typo in commit message " downloading the firmware to DSP"

ujfalusi replied:
Thanks!

ranj063 commented on sound/soc/sof/intel/hda.c:
```diff
@@ -541,8 +541,7 @@ void hda_dsp_dump(struct snd_sof_dev *sdev, u32 flags)
 
-	/* print panic info if FW boot is complete. Otherwise, print the extended ROM status */
-	if (sdev->fw_state == SOF_FW_BOOT_COMPLETE) {
+	if (flags & SOF_DBG_DUMP_REGS) {
```
out of curiosity @ujfalusi , what do you see if you dump regs before FW boot succes?

ujfalusi replied:
I have not made the firmware crash deliberately, but we should have similar information after the boot started as we have after the boot has completed.
I can try to induce some sort of crash, but the best test would be on a real issue.

ujfalusi commented:
Changes since v1:
- Fix typo in commit message pointed out by @ranj063

The aim of this series remain to provide more detailed DSP dump when the panic/crash happens before the firmware is able to send the BOOT_COMPLETE message, but after the firmware itself is started to boot (the firmware is able to send panic message to the host for example).
It _might_ (or might _not_) be a possibility that we could get some extra information **before** the DSP started to execute the firmware, but that must be debugged and validated before we attempt to enable dump at that level.

ujfalusi commented:
I have re-run the PR test via internal CI interface and with the same binaries it is all green, the run ID is 6795.

kv2019i reviewed (commented):
@ujfalusi I welcome cleanups to SOF_DBG_, but the commits messages were quite clear what is the benefit of this change. Is there some new case where we would print the registers where we previously wouldn't?

ujfalusi commented:
> @ujfalusi I welcome cleanups to SOF_DBG_, but the commits messages were quite clear what is the benefit of this change. Is there some new case where we would print the registers where we previously wouldn't?

@kv2019i, yes. We have few reports where we could have more information because the dump is initiated from a panic message from the DSP but before the BOOT_COMPLETE message.
For example this: https://github.com/thesofproject/sof/issues/4538#issuecomment-916858614

ujfalusi commented:
> @ujfalusi I welcome cleanups to SOF_DBG_

@kv2019i, if I were to do a cleanup around the SOF_DBG_ things, than all would be removed, except SOF_DBG_DUMP_REGS and SOF_DBG_DUMP_OPTIONAL, the rest is unused (SOF_DBG_DUMP_MBOX, SOF_DBG_DUMP_TEXT and SOF_DBG_DUMP_PCI) without any hint how they supposed to be used.

I think, this might worth a revisit later.

kv2019i commented:
@ujfalusi wrotE:
> @kv2019i, yes. We have few reports where we could have more information because the dump is initiated from a panic message from the DSP but before the BOOT_COMPLETE message.
> For example this: [thesofproject/sof#4538 (comment)](https://github.com/thesofproject/sof/issues/4538#issuecomment-916858614)

Ack. This would be good to mention in the git commit messages to help upstream consumers understand why we are moving things around.

ujfalusi commented:
> @ujfalusi wrotE:
> 
> > @kv2019i, yes. We have few reports where we could have more information because the dump is initiated from a panic message from the DSP but before the BOOT_COMPLETE message.
> > For example this: [thesofproject/sof#4538 (comment)](https://github.com/thesofproject/sof/issues/4538#issuecomment-916858614)
> 
> Ack. This would be good to mention in the git commit messages to help upstream consumers understand why we are moving things around.

The cover letter for the series is going to mention it as we need all three patches to achieve the said goal, is that OK?
